bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java

```diff
@@ -0,0 +1,276 @@
+        ParadoxCommunicatorFactory factory = new ParadoxCommunicatorFactory(ipAddress, tcpPort, ip150Password,
+                pcPassword);
+        return factory.createCommunicator(panelTypeStr);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Start initialize()...");
+        updateStatus(ThingStatus.UNKNOWN);
+        logger.info("Starting creation of communicator.");
+        config = getConfigAs(ParadoxIP150BridgeConfiguration.class);
+        updateStatus(ThingStatus.OFFLINE);
+
+        scheduler.execute(this::initializeCommunication);
+        logger.debug("Finished initialize().");
+    }
+
+    private void initializeCommunication() {
+        try {
+            communicator = initializeCommunicator();
+            if (communicator.isOnline()) {
+                logger.info("Communicator created successfully. Update Data cache...");
+                updateDataCache(true);
+
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } catch (IOException | InterruptedException | ParadoxBindingException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Error initializing panel handler. Exception: " + e);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        cancelSchedule(refreshCacheUpdateSchedule);
+        communicator.close();
+        super.dispose();
+    }
+
+    private void scheduleRefresh() {
+        logger.debug("Scheduling cache update. Refresh interval: {}s. Starts after: {}s.", config.getRefresh(),
+                INITIAL_SCHEDULE_DELAY);
+        refreshCacheUpdateSchedule = scheduler.scheduleWithFixedDelay(this::updateDataCache, INITIAL_SCHEDULE_DELAY,
+                config.getRefresh(), TimeUnit.SECONDS);
+    }
+
+    private void updateDataCache() {
+        updateDataCache(false);
+    }
+
+    private void updateDataCache(boolean withEpromValues) {
+        try {
+            logger.debug("Refreshing memory map");
+            RawStructuredDataCache cache = RawStructuredDataCache.getInstance();
+
+            boolean isOnline = communicator.isOnline();
+            cache.setIsOnline(isOnline);
+
+            if (isOnline) {
+                communicator.refreshMemoryMap();
+
+                cache.setPanelInfoBytes(communicator.getPanelInfoBytes());
+                cache.setPartitionStateFlags(communicator.readPartitionFlags());
+                cache.setZoneStateFlags(communicator.readZoneStateFlags());
+
+                if (withEpromValues) {
+                    cache.setPartitionLabels(communicator.readPartitionLabels());
+                    cache.setZoneLabels(communicator.readZoneLabels());
+                }
+            }
+
+            announceUpdateToHandlers();
+
+            if (failedReadAttempts > 0) {
+                logger.info("Successfully refreshed memory map after {} failed attempts.", failedReadAttempts);
+                failedReadAttempts = 0;

```

Comment:
Should we update the status here?

Reply:
The status handling is done via the handleSocketReadError method. I would prefer to keep trying without updating thing status until the treshold is reached, otherwise all things start flapping down/up which makes a lot of changes. Not sure if it's good idea. Anyway the things are getting all the data from the last cached result. If something goes really wrong they will get offline after reaching there bridge retry treshold together with the bridge. If it's something temporary I keep the status online. Is that wrong that we have certain amount of time of false behavior saying it's online while it's actually retrying?

Reply:
That sounds good, thanks for the explanation